lib/gitlab/gitlab.go

```diff
@@ -129,3 +132,12 @@ func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
 	}
 	return res, nil
 }
+
+func (c *IDTokenClaims) SetJoinAttrs(in *workloadidentityv1pb.JoinAttrs) proto.Message {
```

Comment:
Need some opinions here - is it weird that this both mutates the passed in struct and returns the data it has injected? I'm not seeing too many other options but I'd like a second opinion in case I'm missing a simpler option.

Roughly - what I'm trying to achieve is:

- Inject the join metadata into workloadidentityv1pb.JoinAttrs for inclusion in the X509 cert as join attributes.
- Continue to be able to place this join metadata into the GoGo structpb.Struct in the audit event, and the Google structpb.Struct in the BotInstanceStatusAuthentication
- Avoid too much duplicated "per join method" code in the lib/auth/join.go handler. Ideally all the join methods can return an individual struct that implements a single method interface that can then be called by code further along to inject/extract this metadata.

Reply:
I guess at a minimum we could probably rename it to `SetAndReturnJoinAttrs()` or something so the behavior is at least not surprising.

That said, the thing that stands out to me is that we have 2 copies of each provider's JWT spec floating around now, one in proto form, and one as the `IDTokenClaims` for a particular provider, with copious amounts of copying between them. I'm guessing the idea is that our `IDTokenClaims` could be a different subset of the provider's claims than what Workload ID wants (or what we want to put in the audit log), but in practice I'm not sure if there's a real distinction between the two, or if there needs to be one?

In other words, could we make the proto definition the source of truth and eliminate `IDTokenClaims`? I think that would simplify the "convert, set, and return" logic to just "set", since we could just copy the full object into the `workloadidentityv1pb.JoinAttrs`. I think `oidc.IDToken.Claims()` deserialization would still work since they resulting `.pb.go` has `json` tags on the fields.

Reply:
> That said, the thing that stands out to me is that we have 2 copies of each provider's JWT spec floating around now, one in proto form, and one as the IDTokenClaims for a particular provider,

So I was mulling over this. It'd be certainly convenient if we could unmarshal directly into this struct, but, I think there's might be a few cases in which we'll want to convert a canonical field from an IdP into a more correct type.

If we take GitLab as an example, they have "ref_protected" and "environment_protected" claims which store a boolean value in string form (e.g "true" or "false"). This wasn't too problematic in our initial use-case where we've been storing these values for the purpose of auditing, but, as we start to look at rule writing/expressions/templating interacting with this data, then, us being able to convert these types for the user so that expressions can be more sane (e.g `join.gitlab.environment_protected && join.gitlab.ref_protected` rather than `join.gitlab.environment_protected == "true" && join.gitlab.ref_protected == "true"`) seems more valuable.

I suppose it's worth considering making `IDTokenClaims` unexported and the validator directly returning a `workloadidentityv1pb.GItLabJoinAttrs`. That would certainly "simplify" - and - in some regards, IDTokenClaims then just becomes an internal detail around how the JWT is unmarshaled and would effectively "not exist".